File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/artists/list/ArtistsListPresenter.java
package com.github.anrimian.musicplayer.ui.library.artists.list;

import com.github.anrimian.musicplayer.domain.business.library.LibraryArtistsInteractor;
import com.github.anrimian.musicplayer.domain.models.artist.Artist;
import com.github.anrimian.musicplayer.domain.models.composition.order.Order;
import com.github.anrimian.musicplayer.domain.utils.TextUtils;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import moxy.InjectViewState;
import moxy.MvpPresenter;

import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.dispose;

@InjectViewState
public class ArtistsListPresenter extends MvpPresenter<ArtistsListView> {

    private final LibraryArtistsInteractor interactor;
    private final ErrorParser errorParser;
    private final Scheduler uiScheduler;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();
    private Disposable artistsDisposable;
    private Disposable changeDisposable;

    private List<Artist> artists = new ArrayList<>();

    @Nullable
    private String searchText;

    public ArtistsListPresenter(LibraryArtistsInteractor interactor,
                                ErrorParser errorParser,
                                Scheduler uiScheduler) {
        this.interactor = interactor;
        this.errorParser = errorParser;
        this.uiScheduler = uiScheduler;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnArtistsList();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onTryAgainLoadCompositionsClicked() {
        subscribeOnArtistsList();
    }

    void onOrderMenuItemClicked() {
        getViewState().showSelectOrderScreen(interactor.getOrder());
    }

    void onOrderSelected(Order order) {
        interactor.setOrder(order);
    }

    void onSearchTextChanged(String text) {
        if (!TextUtils.equals(searchText, text)) {
            searchText = text;
            subscribeOnArtistsList();
        }
    }

    void onNewArtistNameEntered(String name, long artistId) {
        dispose(changeDisposable);
        changeDisposable = interactor.updateArtistName(name, artistId)
                .observeOn(uiScheduler)
                .doOnSubscribe(d -> getViewState().showRenameProgress())
                .doFinally(() -> getViewState().hideRenameProgress())
                .subscribe(() -> {}, this::onDefaultError);
    }

    @Nullable
    String getSearchText() {
        return searchText;
    }

    private void subscribeOnArtistsList() {
        if (artists.isEmpty()) {
            getViewState().showLoading();
        }
        dispose(artistsDisposable, presenterDisposable);
        artistsDisposable = interactor.getArtistsObservable(searchText)
                .observeOn(uiScheduler)
                .subscribe(this::onArtistsReceived, this::onArtistsReceivingError);
        presenterDisposable.add(artistsDisposable);
    }

    private void onArtistsReceivingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showLoadingError(errorCommand);
    }

    private void onArtistsReceived(List<Artist> artists) {
        this.artists = artists;
        getViewState().submitList(artists);
        if (artists.isEmpty()) {
            if (TextUtils.isEmpty(searchText)) {
                getViewState().showEmptyList();
            } else {
                getViewState().showEmptySearchResult();
            }
        } else {
            getViewState().showList();
        }
    }

    private void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showErrorMessage(errorCommand);
    }
}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/artists/list/ArtistsListPresenter.kt
package com.github.anrimian.musicplayer.ui.library.artists.list

import com.github.anrimian.musicplayer.data.utils.rx.RxUtils
import com.github.anrimian.musicplayer.domain.interactors.library.LibraryArtistsInteractor
import com.github.anrimian.musicplayer.domain.models.artist.Artist
import com.github.anrimian.musicplayer.domain.models.order.Order
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.domain.utils.TextUtils
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.Disposable
import java.util.*

class ArtistsListPresenter(private val interactor: LibraryArtistsInteractor,
                           errorParser: ErrorParser,
                           uiScheduler: Scheduler) 
    : AppPresenter<ArtistsListView>(uiScheduler, errorParser) {
    
    private var artistsDisposable: Disposable? = null

    private var artists: List<Artist> = ArrayList()
    
    private var searchText: String? = null

    private var lastEditAction: Completable? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnArtistsList()
    }

    fun onStop(listPosition: ListPosition) {
        interactor.saveListPosition(listPosition)
    }

    fun onTryAgainLoadCompositionsClicked() {
        subscribeOnArtistsList()
    }

    fun onOrderMenuItemClicked() {
        viewState.showSelectOrderScreen(interactor.order)
    }

    fun onOrderSelected(order: Order?) {
        interactor.order = order
    }

    fun onSearchTextChanged(text: String?) {
        if (!TextUtils.equals(searchText, text)) {
            searchText = text
            subscribeOnArtistsList()
        }
    }

    fun onNewArtistNameEntered(name: String?, artistId: Long) {
        lastEditAction = interactor.updateArtistName(name, artistId)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showRenameProgress() }
                .doFinally { viewState.hideRenameProgress() }
        lastEditAction!!.justSubscribe(this::onDefaultError)
    }

    fun onRetryFailedEditActionClicked() {
        if (lastEditAction != null) {
            lastEditAction!!.doFinally { lastEditAction = null }.justSubscribe(this::onDefaultError)
        }
    }

    fun getSearchText() = searchText

    private fun subscribeOnArtistsList() {
        if (artists.isEmpty()) {
            viewState.showLoading()
        }
        RxUtils.dispose(artistsDisposable, presenterDisposable)
        artistsDisposable = interactor.getArtistsObservable(searchText)
                .observeOn(uiScheduler)
                .subscribe(this::onArtistsReceived, this::onArtistsReceivingError)
        presenterDisposable.add(artistsDisposable!!)
    }

    private fun onArtistsReceivingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showLoadingError(errorCommand)
    }

    private fun onArtistsReceived(artists: List<Artist>) {
        val firstReceive = this.artists.isEmpty()

        this.artists = artists
        viewState.submitList(artists)
        if (artists.isEmpty()) {
            if (TextUtils.isEmpty(searchText)) {
                viewState.showEmptyList()
            } else {
                viewState.showEmptySearchResult()
            }
        } else {
            viewState.showList()
            if (firstReceive) {
                val listPosition = interactor.savedListPosition
                if (listPosition != null) {
                    viewState.restoreListPosition(listPosition)
                }
            }
        }
    }

    private fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showErrorMessage(errorCommand)
    }

}